### PR TITLE
Improve the logic of generating running golden config. 

### PR DESCRIPTION
**Pull request by yutongzhang-microsoft**
### Description of PR

In this pr, I improve the logic of generating running golden config. Previously, running golden config is considered as the config before each module running. Now, the running golden config is generated after pre-test or before all the test cases running. 

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205

#### What is the motivation for this PR?
In this pr, I improve the logic of generating running golden config. Previously, running golden config is considered as the config before each module running. Now, the running golden config is generated after pre-test or before all the test cases running. 

#### How did you do it?
- In deploy stage, remove running golden config if it exists. 
- If pre-test runs, generate running golden config after all pre-test cases running. 
- Otherwith, generate running golden config before all test cases running. 

#### How did you verify/test it?
Running test cases with/without running pre-test.

### CODE REVIEW

File: tests/test_pretest.py [hunk 1]
```diff
@@ -278,3 +278,11 @@ def prepare_autonegtest_params(duthosts, fanouthosts):
...
     except IOError as e:
         logger.warning('Unable to create a datafile for autoneg tests: {}. Err: {}'.format(filepath, e))
+
+def test_generate_running_golden_config(duthosts):
+    """
+    Generate running golden config after pre test.
+    """
+    for duthost in duthosts:
+        json_data = json.loads(duthost.shell("sonic-cfggen -d --print-data", verbose=False)['stdout'])
+        duthost.copy(content=json.dumps(json_data, indent=4), dest="/etc/sonic/running_golden_config.json")
```
Comment (wangxin) [inline tests/test_pretest.py:288]: If it is to store golden config on DUT, why not just use command like this:
```
"sonic-cfggen -d --print-data > /etc/sonic/running_golden_config.json"
```

File: tests/conftest.py [hunk 2]
```diff
@@ -1700,10 +1700,20 @@ def core_dump_and_config_check(duthosts, request):
...
+
+            if not golden_config:
+                logger.info("Collecting running golden config before test on {}".format(duthost.hostname))
+                duts_data[duthost.hostname]["pre_running_config"] = \
+                    json.loads(duthost.shell("sonic-cfggen -d --print-data", verbose=False)['stdout'])
+                duthost.copy(content=json.dumps(duts_data[duthost.hostname]["pre_running_config"], indent=4),
+                             dest="/etc/sonic/running_golden_config.json")
+            else:
+                logger.info("Running golden config already exists")
+                duts_data[duthost.hostname]["pre_running_config"] = json.loads(golden_config)
```
Comment (wangxin) [inline tests/conftest.py:1716]: The logic can be simplified a little bit like below:
```
if /etc/sonic/running_golden_config.json not exists:
    dump running config to /etc/sonic/running_golden_config.json

Get pre_running_config from /etc/sonic/running_golden_config.json
```

File: tests/conftest.py [hunk 3]
```diff
@@ -1700,10 +1700,20 @@ def core_dump_and_config_check(duthosts, request):
             else:
                 pre_existing_core_dumps = duthost.shell('ls /var/core/')['stdout'].split()
             duts_data[duthost.hostname]["pre_core_dumps"] = pre_existing_core_dumps
-
-            logger.info("Collecting running config before test on {}".format(duthost.hostname))
-            duts_data[duthost.hostname]["pre_running_config"] = \
-                json.loads(duthost.shell("sonic-cfggen -d --print-data", verbose=False)['stdout'])
+            try:
+                golden_config = duthost.shell("cat /etc/sonic/running_golden_config.json", verbose=False)['stdout']
```
Comment (wangxin) [inline tests/conftest.py:1704]: The ansible module `stat` is able to test if a file exists.

Comment (yutongzhang-microsoft) [reply tests/conftest.py:1704]: done

**Review by wangxin — APPROVED** [review]

**Comment by wangxin** [comment]
@yutongzhang-microsoft There are hard to resolve merge conflicts while cherry-picking to 202012 branch. Can you create a separate PR to the 202012 branch for this change?

**Pull request merged**